hivemind/p2p/p2p_daemon.py

```diff
@@ -73,52 +76,58 @@ def _initialize(self, proc_args: tp.List[str]) -> None:
 
     async def _identify_client(self, delay):
         await asyncio.sleep(delay)
-        try:
-            encoded = await self._client.identify()
-            self.id = encoded[0].to_base58()
-        except Exception:
-            raise
+        encoded = await self._client.identify()
+        self.id = encoded[0].to_base58()
 
     def _assign_daemon_ports(self, host_port=None, daemon_listen_port=None):
         self._host_port, self._daemon_listen_port = host_port, daemon_listen_port
         if host_port is None:
             self._host_port = find_open_port()
-            if daemon_listen_port is None:
-                self._daemon_listen_port = find_open_port()
-                while self._daemon_listen_port == self._host_port:
-                    self._daemon_listen_port = find_open_port()
         if daemon_listen_port is None:
             self._daemon_listen_port = find_open_port()
+            while self._daemon_listen_port == self._host_port:
+                self._daemon_listen_port = find_open_port()
 
     @staticmethod
     async def send_data(data, stream):
         byte_str = pickle.dumps(data)
         request = len(byte_str).to_bytes(P2P.HEADER_LEN, P2P.BYTEORDER) + byte_str
         await stream.send_all(request)
 
-    @staticmethod
-    async def receive_data(stream, max_bytes=(1 << 16)):
-        data = await stream.receive_some(max_bytes)
-        content_length = int.from_bytes(data[:P2P.HEADER_LEN], P2P.BYTEORDER)
-        chunks = [data[P2P.HEADER_LEN:]]
-        curr_length = len(data) - P2P.HEADER_LEN
+    class IncompleteRead(Exception):
+        pass
 
-        while curr_length < content_length:
+    async def _receive_exactly(self, stream, n_bytes, max_bytes=1 << 16):
+        while len(self._buffer) < n_bytes:
             data = await stream.receive_some(max_bytes)
```

Comment:
I suggest to replace `self._buffer` with a local variable `buffer`, remove `max_bytes`, and just write:
```python
data = await stream.receive_some(n_bytes - len(buffer))
```

Then you'll never get more data than needed, so you won't need to save the remainder in `self._buffer`.

Reply:
@skobellev Please note this comment as well :)